### PR TITLE
Table View: Hide/Show Columns - AngularJS

### PR DESCRIPTION
**Pull request by dabeng**
This is Hide/Show Columns feature of angular-datatables which behaves like the recommended PatternFly pattern.

This PR will add colvis dropdown menu to pfToolbar and pfTableView - with Toolbar.

@bleathem @jeff-phillips-18 , would you like to review it?

### CODE REVIEW

File: src/sort/sort.less [hunk 1]
```diff
@@ -1,4 +1,5 @@
 .sort-pf {
+  display: inline-block;
```
Comment (jeff-phillips-18) [inline src/sort/sort.less]: I fear this change. sort-pf is used in many places by many applications in and out of toolbars. Why is this necessary?

File: src/table/tableview/table-view.component.js [hunk 2]
```diff
@@ -3,6 +3,7 @@ angular.module('patternfly.table').component('pfTableView', {
     config: '<?',
     dtOptions: '<?',
     colummns: '<',
+    colvis: '<',
```
Comment (jeff-phillips-18) [inline src/table/tableview/table-view.component.js]: colvis is not intuitive as to what it is used for. Please spell it out as column-visibility

File: src/table/tableview/table-view.component.js [hunk 3]
```diff
@@ -136,6 +138,13 @@ angular.module('patternfly.table').component('pfTableView', {
         selectRowsByChecked();
         //});
       }
+      if (!angular.equals(ctrl.colvis, prevColvis)) {
+        if (ctrl.debug) {
+          $log.debug("  colvis !== prevColvis");
+        }
```
Comment (jeff-phillips-18) [inline src/table/tableview/table-view.component.js]: The debugging should be removed.

File: src/toolbars/toolbar-component.js [hunk 4]
```diff
@@ -1,6 +1,7 @@
 angular.module('patternfly.toolbars').component('pfToolbar', {
   bindings: {
-    config: '='
+    config: '=',
+    onToggleColvis: '&'
```
Comment (jeff-phillips-18) [inline src/toolbars/toolbar-component.js]: colvis is not intuitive as to what it is used for. Please spell it out as onToggleColumnVisibility

File: src/table/tableview/table-view.component.js [hunk 5]
```diff
@@ -67,6 +68,7 @@ angular.module('patternfly.table').component('pfTableView', {
       // Need to deep watch changes in dtOptions and items
```
Comment (jeff-phillips-18) [inline src/table/tableview/table-view.component.js]: Comment update

File: src/table/tableview/table-view.component.js (hunk 2 above)
Comment (jeff-phillips-18) [inline src/table/tableview/table-view.component.js]: I don't think I agree with using colvis to update the visibility of columns. For one thing, how would I set the table up initially? How would I set 2 or more columns to be not visible?

File: src/table/tableview/table-view.component.js [hunk 6]
```diff
@@ -136,6 +138,13 @@ angular.module('patternfly.table').component('pfTableView', {
         selectRowsByChecked();
         //});
       }
+      if (!angular.equals(ctrl.colvis, prevColvis)) {
+        if (ctrl.debug) {
+          $log.debug("  colvis !== prevColvis");
+        }
+        prevColvis = angular.copy(ctrl.colvis);
+        ctrl.dtColumnDefs[ctrl.colvis.index].visible = ctrl.colvis.visible;
```
Comment (jeff-phillips-18) [inline src/table/tableview/table-view.component.js]: Need to check if the column defs actually contain this field.

File: src/table/tableview/table-view.component.js (hunk 5 above)
Comment (jeff-phillips-18) [inline src/table/tableview/table-view.component.js]: onToggleColvis should be in the config object to be consistent with other callback functions.

File: src/toolbars/toolbars.less [hunk 7]
```diff
@@ -29,3 +29,22 @@
 .toolbar-pf-actions.no-filter-results {
   margin-bottom: 10px;
 }
+.dropdown-colvis-pf {
+  margin-left: 15px;
+  .dropdown-menu {
+    > li {
+      padding: 5px 10px;
+      &:hover {
+        background-color: #def3ff;
```
Comment (jeff-phillips-18) [inline src/toolbars/toolbars.less]: This should be a color palette variable.

File: src/toolbars/toolbars.less [hunk 8]
```diff
@@ -29,3 +29,22 @@
...
+      }
+      > input {
+        margin-top: 0;
+        margin-right: 10px;
```
Comment (jeff-phillips-18) [inline src/toolbars/toolbars.less]: alphabetical order please.

File: src/toolbars/toolbars.less [hunk 9]
```diff
@@ -29,3 +29,22 @@
...
+        vertical-align: middle;
+      }
+      > label {
+        margin-bottom: 0;
+      }
+    }
+  }
+}
```
Comment (jeff-phillips-18) [inline src/toolbars/toolbars.less]: Please add a new line.

**Comment by jeff-phillips-18** [comment]
Overall, I don't really agree with using the first item in the list of items to determine the columns in the table view component. Seems to me that dtColumnDefs should be a binding to allow the application to set it up how they want. What happens when the items are empty when the table is initialized and then items are filled in later? While this is not really part of this PR, trying to set the initial column visibilities may cause you to rethink this.

**Comment by jeff-phillips-18** [comment]
Need ng-doc updates to include new bindings for both toolbar and table component.

**Comment by jeff-phillips-18** [comment]
Need updates to table component unit test.

**Comment by jeff-phillips-18** [comment]
Also, while we are still in alpha, it might be a good time to put the table view component with list view and card view components in src/views.

File: src/sort/sort.less (hunk 1 above)
Comment (dabeng) [inline src/sort/sort.less]: Without this setting, Sort control group will look like the following screenshot.
screen shot 2017-03-07 at 11 00 47 am

File: src/table/tableview/table-view.component.js (hunk 2 above)
Comment (dabeng) [inline src/table/tableview/table-view.component.js]: (1) colvis is a terminology of DataTables plugin. From my point of view, we should respect the original ideas of authorship and usage habit of veteran datatables users. FYI, https://datatables.net/reference/button/colvis.

Comment (dabeng) [inline src/table/tableview/table-view.component.js]: 2.
(1)  The task "hide some columns initially" has nothing to do with colvis feature included in toolbar. No matter toolbar provides colvis or not, if users want to hide columns on initial timing, pfTableview should supply the feasible solution to handle it.

(2) However, current code use the hybrid mode to render datatables -- pre-populated html dom table drived by angular + core api of jquery datatables. Such an approach directly causes the lightweight solution as shown below not working.
```js
      $scope.dtOptions = {
        paginationType: 'full',
        displayLength: 10,
        dom: "tp",
        columnDefs: [                      //
          { targets: [3], visible: false },//
        ]                                  //
      };
```
(3) FYI, the following is just a temporary solution.
```js
      $scope.dtOptions = {
        paginationType: 'full',
        displayLength: 10,
        dom: "tp",
        initComplete: function( settings, json ) {//
          this.api().column(3).visible(false);    //
        }                                         //
      };
```

File: src/table/tableview/table-view.component.js (hunk 5 above)
Comment (dabeng) [inline src/table/tableview/table-view.component.js]: There is no onToggleColvis in table-view.component.js file. It's in toolbar-component.js file. Additionally, I just followed the inputs/outputs styles described in [angular guide](https://docs.angularjs.org/guide/component#component-based-application-architecture). Is it necessary to customize the callback mechanism by ourselves ?

File: src/table/tableview/table-view.component.js (hunk 3 above)
Comment (dabeng) [inline src/table/tableview/table-view.component.js]: got it

File: src/table/tableview/table-view.component.js (hunk 6 above)
Comment (dabeng) [inline src/table/tableview/table-view.component.js]: What's the "this field"?

**Comment by dabeng** [comment]
With respect, jquery datatables, angular-datatables and pfTableView are in three different levels of abstraction. But it seems that our current code do a lot of work which jquery datatables or angular-datatables should take on. This is exactly the matter that I worry most because this situation will probably become the first big hindrance we add new features or fix bugs for pfTableView in the future.

**Comment by dtaylor113** [comment]
> Also, while we are still in alpha, it might be a good time to put the table view component with list view and card view components in src/views.

Hi Jeff, 'table view' requires dtables.net and jQuery so I intentionally set it in it's own module so that users could have list and card view without having to also pull in dtables/jQuery.

File: src/table/tableview/table-view.component.js (hunk 2 above)
Comment (dtaylor113) [inline src/table/tableview/table-view.component.js]: > (2) However, current code use the hybrid mode to render datatables -- pre-populated html dom table drived by angular + core api of jquery datatables. Such an approach directly causes the lightweight solution as shown below not working.

Why doesn't this work?  dtOptions are passed directly to angular-datatables, unchanged by our code.  If angular-datatables supports dtOptions as you have outlined, then it should work.

Comment (dtaylor113) [inline src/table/tableview/table-view.component.js]: http://l-lin.github.io/angular-datatables/archives/#!/withColVis

File: src/toolbars/examples/toolbar.js [hunk 10]
```diff
@@ -50,7 +50,7 @@
   <file name="index.html">
     <div ng-controller="ViewCtrl" class="row example-container">
       <div class="col-md-12">
-        <pf-toolbar id="exampleToolbar" config="toolbarConfig">
+        <pf-toolbar id="exampleToolbar" config="toolbarConfig" on-toggle-colvis="toggleColvis(column)">
          <actions>
```
Comment (dtaylor113) [inline src/toolbars/examples/toolbar.js]: Up until now pfToolbar has been independent of any 'view' (list, card).  It seems wrong to add a table-view specific attribute here.  Especially,  since (and I agree with @jeff-phillips-18) it's name is very cryptic.   Could we rename it to make it clearer?  How about 'toggleTableColsVisibility'.  I know 'colvis' is the name the plug-in uses, but it has been depreciated, and soon will be forgotten.

File: src/toolbars/examples/toolbar.js [hunk 11]
```diff
@@ -119,6 +119,7 @@
       <div class="col-md-12" ng-show="viewType == 'tableView'">
         <pf-table-view config="tableConfig"
                        colummns="colummns"
+                       colvis="colvis"
                        items="items"
```
Comment (dtaylor113) [inline src/toolbars/examples/toolbar.js]: Couldn't we add the properties of 'colvis' to the already in use 'columns' object? Seems logical to add 'index' and 'visible' to columns obj. which already has other column attributes.

File: src/table/tableview/table-view.component.js (hunk 2 above)
Comment (dabeng) [inline src/table/tableview/table-view.component.js]: Because of the following sentence:
```js
 <td ng-repeat="(key, value) in item" ng-if="$ctrl.isColItemFld(key)">{{ value }}</td>
```

File: src/toolbars/examples/toolbar.js (hunk 11 above)
Comment (dabeng) [inline src/toolbars/examples/toolbar.js]: First, $scope.colummns is special for defining table head. Second, "colvis" is just an intermediate variable which is in charge of telling pfTableView which column's visibility is toggled. Any time, there is only one colvis variable.

File: src/toolbars/examples/toolbar.js (hunk 10 above)
Comment (dabeng) [inline src/toolbars/examples/toolbar.js]: Hi @dtaylor113 , I just followed the inputs/outputs styles described in [angular guide](https://docs.angularjs.org/guide/component#component-based-application-architecture). Here, we expose the "column" variable to its parent controller scope in order to communicate with sibling component pfTableView.

**Comment by bleathem** [comment]
+1 for using `column-visibility` over `colvis`.  I imagine the majority of our users will come to using the table view without having previously used datatables.net, and so providing a more descriptive name will make it more clear how to consume the API.  We've also decided not to use the `buttons` library, as it doesn't allow for sufficient customisation to match PatternFly patterns.

**Comment by jeff-phillips-18** [comment]
@dabeng Sorry, I am consumed with other work at the moment. I will get back to this when I free up a bit.

**Comment by dabeng** [comment]
Thanks a lot 😊

**Comment by bleathem** [comment]
Apparently I closed this when I deleted the 4.0-dev branch.

@dabeng can you please rebase this PR against master (master is now 4.0) and reopen it?
@jeff-phillips-18 are we prepared to resume our review of this PR?
